packages/next/next-server/server/next-server.ts

```diff
@@ -253,13 +253,11 @@ export default class Server {
 
     const { basePath } = this.nextConfig
 
-    // if basePath is set require it be present
-    if (basePath && !req.url!.startsWith(basePath)) {
-      return this.render404(req, res, parsedUrl)
-    } else {
-      // If replace ends up replacing the full url it'll be `undefined`, meaning we have to default it to `/`
-      parsedUrl.pathname = parsedUrl.pathname!.replace(basePath, '') || '/'
-      req.url = req.url!.replace(basePath, '')
+    if (basePath && req.url?.startsWith(basePath)) {
+      // store original URL to allow checking if basePath was
+      // provided or not
+      ;(req as any)._nextHadBasePath = true
+      req.url = req.url!.replace(basePath, '') || '/'
```

Comment:
@ijjk @timneutkens what about adding `req.baseUrl = basePath;` to [express req object](https://expressjs.com/en/api.html#req.baseUrl) in this code block?

I have a problem that I cannot make a proper `redirect` in `getInitialProps()` with `basePath`. I have no ability to read it from `ctx` or `ctx.req` object on the server-side for proper redirect URL construction.